src/lib/operators/table_scan.cpp

```diff
@@ -239,10 +265,38 @@ std::shared_ptr<AbstractExpression> TableScan::_resolve_uncorrelated_subqueries(
         subquery_result = AllTypeVariant{expression_result->value(0)};
       }
     });
-    argument = std::make_shared<ValueExpression>(std::move(subquery_result));
+    materialized_arguments.emplace_back(std::make_shared<ValueExpression>(std::move(subquery_result)));
+
+    // Deregister, because we obtained the subquery result and no longer need the subquery plan.
+    subquery->pqp->deregister_consumer();
+    computed_subqueries_count++;
```

Comment:
(Why) does this work if two scans share the same uncorrelated subquery? Given that this is part of `_on_execute`, wouldn't we release the result that we later want to reuse?

Reply:
> (Why) does this work if two scans share the same uncorrelated subquery?

You mean two TableScans sharing the same uncorrelated subquery? 

**Given:** TableScan1 & TableScan2 have predicates with arguments referring to the same uncorrelated subquery X

1. TableScan1 & TableScan2 both register at X → `_consumer_count = 2`
2.  TableScan1 resolves X & deregisters. Results remain cached because X still has `_consumer_count = 1`.
3. TableScan2 resolves X, reuses the result and deregisters → `_consumer_count = 0`
4. X calls `_clear_output`

TableScan1 & TableScan2 still have the results of X, because they were cached inside ValueExpressions.

```cpp
new_arguments.emplace_back(std::make_shared<ValueExpression>(std::move(subquery_result)));
```


> Given that this is part of _on_execute, wouldn't we release the result that we later want to reuse?

Uncorrelated subqueries are resolved and deregistered after `TableScan::create_impl` was called.

Reply:
Ah, right. The beauty of registering in the constructor. At which point, however, do we register as the consumer for a subquery within the subquery? `find_pqp_subquery_expressions` does not cover nested queries, correct? Do we have a test case for that?

Reply:
> find_pqp_subquery_expressions does not cover nested queries, correct? 

Correct. 

>  At which point, however, do we register as the consumer for a subquery within the subquery?

Subqueries should take care of that themselves because they use the same TableScan/Projection implementations

Reply:
Ah, right. And because those operators are already created, they are properly registered as consumers, correct? Could you maybe expand the comment in abstract_operator.hpp to discuss the handling of nested subqueries?

Reply:
> And because those operators are already created, they are properly registered as consumers, correct? 

Correct.

> Could you maybe expand the comment in abstract_operator.hpp to discuss the handling of nested subqueries?

Noted.

Reply:
From the Doc:
```
 *  WARNING on handling Subqueries:
 *   This abstract class handles consumer registration/deregistration for input operators only. Operators that consume
 *   subqueries, such as TableScan and Projection, have to register and deregister as consumers for those manually.
 *   It is crucial to call register_consumer() from the constructor, before the execution starts, to prevent subquery
 *   results from being cleared too early. Otherwise, operators may need to re-execute, which is illegal.
 ```

Reply:
It is technically in there, but I am not sure if someone who cursorily looks at it will understand the implications. Maybe we can give an example?

Reply:
I could describe how the TableScan or Projection handles this in detail.

Reply:
That would be great. I am sure future readers will be grateful.